### PR TITLE
feature: Add support for SerializableProgram abstraction to Device interface

### PR DESCRIPTION
**Pull request by rmshaffer**
*Issue #, if available:*

*Description of changes:*
The scenario is that an external library like [AutoQASM](https://github.com/amazon-braket/autoqasm) wants to create program objects which can be executed on Braket devices, including local simulators and AWS devices, but also wants the freedom to return (simulated) results in a format that does not necessarily match the Braket schema for gate model task results.

This change adds a minimal abstraction, `SerializableProgram`, which represents a program that can be serialized to OpenQASM. It adds support for this in the Device class so that these programs can be run on Braket, while also supporting a local simulator which wants to return a `GateModelQuantumTaskResult` directly rather than being forced to match the schema of `GateModelTaskResult`.

For example implementations of this abstraction, see:
- https://github.com/amazon-braket/autoqasm/blob/7e4441dd6cc42c1ee01b52d86e1ee7a04fa15b40/src/autoqasm/program/program.py#L121-L154
- https://github.com/amazon-braket/autoqasm/blob/7e4441dd6cc42c1ee01b52d86e1ee7a04fa15b40/src/autoqasm/program/program.py#L227-L261

*Testing done:*
Tests added to verify new code and maintain 100% code coverage.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by laurencap — DISMISSED** [review]

**Comment by yitchen-tim** [comment]
>  It adds support for this in the Device class so that these programs can be run on Braket, while also supporting a local simulator which wants to return a GateModelQuantumTaskResult directly rather than being forced to match the schema of GateModelTaskResult.

If the vision is to support any simulator that can execute OpenQASM, do you think it's worth creating a new result class instead of GateModelQuantumTaskResult? Currently, the MCM simulator writes results in the `measurements` field of `GateModelQuantumTaskResult`. Although `GateModelQuantumTaskResult` does not validate the field type, it does have type hint and doc that does not agree with the value written into the `measurements` field.

**Comment by rmshaffer** [comment]
> If the vision is to support any simulator that can execute OpenQASM, do you think it's worth creating a new result class instead of GateModelQuantumTaskResult? Currently, the MCM simulator writes results in the `measurements` field of `GateModelQuantumTaskResult`. Although `GateModelQuantumTaskResult` does not validate the field type, it does have type hint and doc that does not agree with the value written into the `measurements` field.

You're right, using the `measurements` field for arbitrary `output` values is not really intended. To support arbitrary OpenQASM programs with `output` variables, eventually we would need a task result object which is more generic than the current `GateModelTaskResult` and `GateModelQuantumTaskResult`, both of which assume that the result of a task is a single bitstring. However, doing this correctly would involve defining a new task result schema (in the [schemas repo](https://github.com/amazon-braket/amazon-braket-schemas-python)) which I believe is beyond the scope of this PR.

One possible option for now would be to define some kind of generic `QuantumTaskResult` object which does not have any particular schema. But this feels like it would be a temporary solution and I'd be hesitant to add that now. Curious to get other thoughts on this.

**Review by krneta — APPROVED** [review]

File: src/braket/circuits/serialization.py [hunk 1]
```diff
@@ -32,6 +33,31 @@ class QubitReferenceType(str, Enum):
...
+        self,
+        ir_type: IRType = IRType.OPENQASM,
+        build_if_necessary: bool = False,
+    ) -> str:
+        """Serializes the program into an intermediate representation.
+
+        Args:
+            ir_type (IRType): The IRType to use for converting the program to its
+                IR representation. Defaults to IRType.OPENQASM.
+            build_if_necessary (bool): Whether to allow the program to be implicitly
```
Comment (krneta) [inline src/braket/circuits/serialization.py:48]: This is a little weird. Why do we need it?

Comment (rmshaffer) [reply src/braket/circuits/serialization.py]: Agreed, this arg doesn't belong here. Removing it.

**Review by krneta — APPROVED** [review]

**Pull request merged**